plutarch-docs/src/Typeclasses/PLiftable.md

```diff
@@ -0,0 +1,231 @@
+another for its Haskell-level equivalent. Much as with `DeriveBuiltinPLiftable`,
+we use `DeriveDataPLiftable` as so, substituting `a` and `h` appropriately:
+
+```haskell
+deriving via (DeriveDataPLiftable a h) instance PLiftable a
+```
+
+There are more stringent requirements of `a` and `h` for this to work. Aside
+from `a` being an instance of `PlutusType`, we also must ensure that:
+
+* `PInner a` is `PData` (namely, construction and matching is actually carried
+  out in a computation involving `Data`)
+* `h` is an instance of both `ToData` and `FromData` (namely, it has a `Data`
+  encoding that we can decode from and encode into)
+
+### Manual derivation
+
+In the (unlikely) case that your type fits neither of the above, you will have
+to write the instance manually. Given how unusual this situation is, we can't
+really give any general guidance as to how this should be done. Instead, we
+recommend examining the definitions of `PLiftable`, as well as the two
+derivation helpers described above, in the source code. Alternatively, reach out
+to us: we might be able to advise better.
+
+### Ensuring your instance is correct
+
+Any instance of `PLiftable` must obey the following laws:
+
+```
+1. fromPlutarch . toPlutarch = Right
+2. fmap toPlutarch . fromPlutarch = Right
+```
+
+If you use either of `DeriveBuiltinPLiftable` or `DeriveDataPLiftable` with
+`via` derivation, these laws will automatically be satisfied. If you write an
+instance manually, you will have to ensure this yourself. We provide a helper
+for testing that these laws hold in `plutarch-testlib`, in the
+`Plutarch.Test.Laws` module, called `checkPLiftableLaws`, which uses QuickCheck
+to verify that the laws are maintained.
+
+One important caveat to any definition of `PLiftable` instances, manual or not:
+ensure that the Haskell-level equivalent that you declare is genuine. While
+there is nothing stopping you from defining something like `PLiftable PNatural`
+with `AsHaskell PNatural = Text`, this is clearly not sensible, and we cannot
+check this. Your intent is taken at its word: the only checks are that what you
+want is not literally impossible. Keep this in mind when defining your
+instances.
+
+## Using an instance
+
+To simplify the use of `PLiftable`, we provide two functions:
+
+```haskell
+pconstant :: forall (a :: S -> Type) (s :: S) .
+    PLiftable a =>
+    AsHaskell a -> 
+    Term s a
+
+plift :: forall (a :: S -> Type) .
+    PLiftable a => 
+    (forall (s :: S) . Term s a) -> 
+    AsHaskell a
+```
+
+The type signatures more-or-less speak for themselves: `pconstant` is the
+Haskell-to-Plutarch direction, while `plift` is the Plutarch-to-Haskell
+direction. There are three minor caveats to their use:
+
+* `pconstant` is technically ambiguous, as many different Plutarch types can
+  share the same choice for `AsHaskell`. Thus, you may need to use type
+  arguments when calling `pconstant` to avoid ambiguity errors from the
```

Comment:
Show something like how there can be both
`PLiftable PInteger` and `PLiftable (PAsData PInteger)` both from Haskell `Integer` in which case user would need to `pconstant @(PAsData PInteger)` or `pconstant @PInteger`